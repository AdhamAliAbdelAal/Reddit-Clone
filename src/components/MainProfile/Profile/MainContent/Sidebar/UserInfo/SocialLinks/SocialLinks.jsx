import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import IconButton from '@mui/material/IconButton';
import CloseIcon from '@mui/icons-material/Close';
// import TextField from '@mui/material/TextField';
import { useState } from 'react';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import { Typography } from '@mui/material';
import allSocialLinks from './server';
import {
  InputBox, PlatformIcon, SaveBtn, Text,
} from './styles';

function SocialLinks({ onClose }) {
  const [links] = allSocialLinks();
  const [openDialog2, setDialog2] = useState(false);
  const [platform, setPlatform] = useState(null);
  const [condition, setCondition] = useState(false);
  const [text, setText] = useState('');
  // const [error, setError] = useState(false);
  // let msg = ' ';

  const handleOpenPlatform = (link) => {
    setDialog2(true);
    setPlatform(link);
  };

  const handleClose = () => {
    setDialog2(false);
  };

  const handleCheck = (e) => {
    if (e.target.value.trim().length >= 1 && e.target.value.trim().indexOf('@') !== 0) {
      e.target.value = `@${e.target.value.trim()}`;
    }
    if (e.target.value.trim().length > 0) setCondition(true);
    else setCondition(false);
    setText(e.target.value.trim());
  };

  const checkSubmission = () => {
    if (text.includes(' ')) {
      // setError(true);
      // msg = 'user name is not valid';
    }
  };

  return (
    !openDialog2 ? (
      <>
        <DialogTitle sx={{
          m: 0, p: 2, display: 'flex', justifyContent: 'center', fontWeight: 700,
        }}
        >
          Add Social Link
          {onClose ? (
            <IconButton
              aria-label="close"
              onClick={onClose}
              sx={{
                position: 'absolute',
                right: 8,
                top: 8,
                color: (theme) => theme.palette.grey[500],
              }}
            >
              <CloseIcon />
            </IconButton>
          ) : null}
        </DialogTitle>
        <DialogContent
          dividers
          sx={{
            display: 'flex',
            flexDirection: 'row',
            flexWrap: 'wrap',
          }}
        >
          {links?.map((link, index) => (
            <Text key={`${index + 0}`} onClick={() => handleOpenPlatform(link)}>
              <PlatformIcon src={link?.icon} />
              {link?.type}
            </Text>
          )) }
        </DialogContent>
      </>
    ) : (
      <>
        <DialogTitle sx={{
          m: 0, p: 2, display: 'flex', justifyContent: 'space-between',
        }}
        >
          <IconButton
            onClick={handleClose}
            sx={{
              color: '#878a8c',
            }}
          >
            <ArrowBackIcon />
          </IconButton>
          <Typography
            variant="h6"
            sx={{
              display: 'flex', alignItems: 'center',
            }}
          >
            Add Social Link
          </Typography>
          <SaveBtn variant="contained" condition={condition.toString()} onClick={() => checkSubmission()}>Save</SaveBtn>
        </DialogTitle>
        <DialogContent dividers>
          <Text>
            <PlatformIcon src={platform?.icon} />
            {platform?.type}
          </Text>
          <InputBox type="text" placeholder={platform?.placeholderLink} onChange={(e) => { handleCheck(e); }} />
        </DialogContent>
        {/* {error && <Typography sx={{ color: 'red' }} variant="string">{msg}</Typography>} */}
      </>
    )
  );
}

export default SocialLinks;
