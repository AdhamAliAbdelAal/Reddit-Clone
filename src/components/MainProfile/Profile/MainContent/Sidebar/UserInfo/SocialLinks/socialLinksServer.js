import { useEffect } from 'react';
import useFetch from '../../../../../../../hooks/useFetch';

const allSocialLinks = () => {
  const [data, dataError, statusCode] = useFetch('/users/social_links');

  useEffect(() => {
    if (statusCode === 401) {
      window.location.pathname = 'login';
    }
  }, [data, dataError, statusCode]);

  return [data?.socialLinks];
};

export default allSocialLinks;
