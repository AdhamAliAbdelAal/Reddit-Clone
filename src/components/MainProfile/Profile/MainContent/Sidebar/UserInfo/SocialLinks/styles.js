import {
  Button,
  styled,
} from '@mui/material';

export const SaveBtn = styled(Button)(({ condition }) => ({
  borderRadius: 20,
  textTransform: 'unset',
  fontWeight: 700,
  height: 32,
  boxShadow: 'none',
  '&:hover': {
    boxShadow: 'none',

  },
  ...((condition === 'false') && {
    cursor: 'not-allowed',
    color: '#ffffff80',
    filter: 'grayscale(1)',
  }),
}));

export const InputBox = styled('input')(() => ({
  height: 36,
  width: '100%',
  marginTop: 5,
  marginBottom: 10,
  fontSize: 14,
  border: '1px solid #edeff1',
  borderRadius: 5,
  textIndent: 10,
  '&:focus': {
    borderColor: '#0079d3',
    outline: 'none',
  },
}));

export const Text = styled('span')(() => ({
  height: 40,
  width: 'fit-content',
  fontSize: 12,
  fontWeight: 700,
  backgroundColor: '#edeff1',
  padding: '10px 12px',
  borderRadius: 9999,
  color: 'black',
  cursor: 'pointer',
  marginBottom: 8,
  marginRight: 4,
  textAlign: 'center',
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
}));

export const PlatformIcon = styled('img')(() => ({
  height: 16,
  width: 16,
  marginRight: 5,
}));
