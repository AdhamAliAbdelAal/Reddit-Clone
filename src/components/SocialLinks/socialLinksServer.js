import { useRef, useState, useEffect } from 'react';

import useFetch from '../../hooks/useFetch';
import axios from '../../services/instance';

export const allSocialLinks = () => {
  const [data, dataError, statusCode] = useFetch('users/social_links');

  useEffect(() => {
    if (statusCode === 401) {
      window.location.pathname = 'login';
    }
  }, [data, dataError, statusCode]);

  return [data?.socialLinks];
};

export const postSocialLink = (dataToSend) => {
  const [userLink, displayText, socialId] = dataToSend;
  const [data, setData] = useState(null);
  const isFirstRun = useRef(true);

  useEffect(() => {
    if (isFirstRun.current) {
      isFirstRun.current = false;
      return;
    }
    axios.post('users/social_links/sabry', {
      userLink,
      displayText,
      socialId,
    }).then((response) => {
      if (response.status === 401) {
        window.location.pathname = 'login';
      }
      console.log(response);
      setData(response);
    }).catch((error) => {
      console.log(error);
    });
  }, [dataToSend]);
  return data;
};
