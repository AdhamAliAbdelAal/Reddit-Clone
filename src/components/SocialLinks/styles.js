import {
  Button,
  styled,
} from '@mui/material';

export const SaveBtn = styled(Button)(({ condition }) => ({
  borderRadius: 20,
  textTransform: 'unset',
  fontWeight: 700,
  height: 32,
  boxShadow: 'none',
  '&:hover': {
    boxShadow: 'none',

  },
  ...((condition === 'false') && {
    cursor: 'not-allowed',
    color: '#ffffff80',
    filter: 'grayscale(1)',
  }),
}));

export const InputBox = styled('input')(() => ({
  height: 36,
  width: '100%',
  marginTop: 5,
  marginBottom: 10,
  fontSize: 14,
  border: '1px solid #edeff1',
  borderRadius: 5,
  textIndent: 10,
  '&:focus': {
    borderColor: '#0079d3',
    outline: 'none',
  },
}));
