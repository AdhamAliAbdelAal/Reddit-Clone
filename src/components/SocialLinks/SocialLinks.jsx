import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import IconButton from '@mui/material/IconButton';
import CloseIcon from '@mui/icons-material/Close';
import { useState, useEffect } from 'react';
import ArrowBackIcon from '@mui/icons-material/ArrowBack';
import { Typography } from '@mui/material';
import { allSocialLinks, postSocialLink } from './socialLinksServer';
import {
  InputBox, SaveBtn, Text, PlatformIcon,
} from './styles';

function SocialLinks({
  onClose, settings, link,
}) {
  const [links] = allSocialLinks();
  const [openDialog2, setDialog2] = useState(false);
  const [platform, setPlatform] = useState(null);
  const [condition, setCondition] = useState(false);
  const [text, setText] = useState('');
  const [error, setError] = useState(false);
  const [msg, setMsg] = useState('');
  const [url, setUrl] = useState(false);
  const [edit, setEdit] = useState(true);
  const [title, setTitle] = useState('');
  const [dataToSend, setDataToSend] = useState([]);

  postSocialLink(dataToSend);
  useEffect(() => {
    if (settings) {
      setDialog2(true);
      setPlatform(link?.social);
      if (link?.social.placeholderLink.includes('https://')) {
        setText(link?.userLink);
        setTitle(link?.displayText);
      }
      setUrl(link?.social.placeholderLink.includes('https://'));
    }
  }, [settings]);

  const handleOpenPlatform = (link) => {
    setDialog2(true);
    setPlatform(link);
    setEdit(true);
    setUrl(link?.placeholderLink.includes('https://'));
  };

  const handleClose = () => {
    setError(false);
    setText('');
    setTitle('');
    if (settings) {
      onClose();
    } else {
      setDialog2(false);
    }
  };

  // const handleCheck = (e) => {
  //   if (platform?.placeholderLink[0] === '@' && e.target.value.trim().length >= 1 && e.target.value.trim().indexOf('@') !== 0) {
  //     e.target.value = `@${e.target.value.trim()}`;
  //   } else if (platform?.placeholderLink[0] === 'h' && e.target.value.trim().length >= 1 && !'https://'.includes(e.target.value.trim().substring(0, 6))) {
  //     e.target.value = `https://${e.target.value.trim()}`;
  //   }
  //   if (e.target.value.trim().length > 0) setCondition(true);
  //   else setCondition(false);
  //   setLink(e.target.value.trim());
  // };

  const checkSubmission = () => {
    if (url && (text.includes(' ') || text.length <= 1)) {
      setError(true);
      setMsg('Username is not valid');
    } else if (url && (!text.includes(platform?.baseLink) || text === platform?.baseLink)) {
      setError(true);
      setMsg('Invalid URL');
    } else {
      setError(false);
      setDataToSend([text, title, platform?.type]);
      onClose();
      setDialog2(false);
      setText('');
      setTitle('');
    }
  };
  const handelchange = (e) => {
    if (!settings && edit) {
      setText(platform.placeholderLink + e.target.value.trim());
      setEdit(false);
    } else {
      setText(e.target.value.trim());
    }
    if (e.target.value.trim().length > 0) { setCondition(true); } else { setCondition(false); }
  };
  return (
    !openDialog2 ? (
      <>
        <DialogTitle sx={{
          m: 0, p: 2, display: 'flex', justifyContent: 'center', fontWeight: 700,
        }}
        >
          Add Social Link
          {onClose ? (
            <IconButton
              aria-label="close"
              onClick={onClose}
              sx={{
                position: 'absolute',
                right: 8,
                top: 8,
                color: (theme) => theme.palette.grey[500],
              }}
            >
              <CloseIcon />
            </IconButton>
          ) : null}
        </DialogTitle>
        <DialogContent
          dividers
          sx={{
            display: 'flex',
            flexDirection: 'row',
            flexWrap: 'wrap',
          }}
        >
          {links?.map((link, index) => (
            <Text key={`${index + 0}`} onClick={() => handleOpenPlatform(link)}>
              <PlatformIcon src={link?.icon} />
              {link?.type}
            </Text>
          )) }
        </DialogContent>
      </>
    ) : (
      <>
        <DialogTitle sx={{
          m: 0, p: 2, display: 'flex', justifyContent: 'space-between',
        }}
        >
          <IconButton
            onClick={handleClose}
            sx={{
              color: '#878a8c',
            }}
          >
            <ArrowBackIcon />
          </IconButton>
          <Typography
            variant="h6"
            sx={{
              display: 'flex', alignItems: 'center',
            }}
          >
            Add Social Link
          </Typography>
          <SaveBtn variant="contained" condition={condition.toString()} onClick={() => checkSubmission()}>Save</SaveBtn>
        </DialogTitle>
        <DialogContent dividers>
          <Text>
            <PlatformIcon src={platform?.icon} />
            {platform?.type}
          </Text>
          { url
              && <InputBox type="text" value={title} placeholder="Display Text" onChange={(e) => { setTitle(e.target.value); }} />}
          <InputBox type="text" value={text} placeholder={platform?.placeholderLink} onChange={(e) => { handelchange(e); }} />
          {error ? <Typography sx={{ color: 'red', marginLeft: 1 }} variant="caption">{msg}</Typography> : null}
        </DialogContent>
      </>
    )

  );
}

export default SocialLinks;
