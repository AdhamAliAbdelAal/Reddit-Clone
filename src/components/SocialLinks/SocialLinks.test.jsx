import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import SocialLinks from './SocialLinks';
// test snapshot
test('test snapshot', async () => {
  const tree = renderer.create(
    <Router>
      <SocialLinks />
    </Router>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
