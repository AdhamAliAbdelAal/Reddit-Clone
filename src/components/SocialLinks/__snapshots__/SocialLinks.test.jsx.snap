// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test snapshot 1`] = `
Array [
  <h2
    className="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-g8doz5-MuiTypography-root-MuiDialogTitle-root"
  >
    Add Social Link
  </h2>,
  <div
    className="MuiDialogContent-root MuiDialogContent-dividers css-1uhmgr5-MuiDialogContent-root"
  />,
]
`;
